File: web/src/lib/clue.ts
interface NumberClue {
  type: "number";
  i: number;
  j: number;
  value: number;
}

interface QuestionMarkClue {
  type: "questionMark";
  i: number;
  j: number;
}

export type Clue = NumberClue | QuestionMarkClue;

File: solver14mv/src/solver14mv/solver/Clue.scala
package solver14mv.solver

enum Clue {
  case None
  case QuestionMark
  case Vanilla(value: Int)
  case Multiple(value: Int)
  case Liar(value: Int)
  case Wall(value: List[Int])
  case Negation(value: Int)
  case Cross(value: Int)
}
